packages/text-editor/src/components/extension/fileExt.ts

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * @public
+ */
+export interface FileOptions extends FileNodeOptions {
+  attachFile?: FileAttachFunction
+  reportNode?: (id: string, node: ProseMirrorNode) => void
+}
+
+/**
+ * @public
+ */
+export const inputRegex = /(?:^|\s)(!\[(.+|:?)]\((\S+)(?:(?:\s+)["'](\S+)["'])?\))$/
+
+/**
+ * @public
+ */
+export const FileExtension = FileNode.extend<FileOptions>({
+  addOptions () {
+    return {
+      inline: true,
+      HTMLAttributes: {}
+    }
+  },
+
+  parseHTML () {
+    return [
+      {
+        tag: `div[data-type="${this.name}"]`
+      },
+      {
+        tag: 'div[file-name]'
+      },
+      {
+        tag: 'div[file-size]'
+      },
+      {
+        tag: 'div[file-type]'
+      },
+      {
+        tag: 'div[src]'
+      }
+    ]
+  },
+
+  priority: 1100,
+
+  renderHTML ({ node, HTMLAttributes }) {
+    const nodeAttributes = {
+      class: 'text-editor-file-container',
+      'data-type': this.name
+    }
+
+    const id = HTMLAttributes['file-id']
+    const fileName = HTMLAttributes['file-name']
+    const size = HTMLAttributes['file-size']
+    const fileType = HTMLAttributes['file-type']
+    const linkAttributes = {
+      class: 'file-name',
+      href: node.attrs.src,
+      type: fileType,
+      download: fileName,
+      target: '_blank'
+    }
+    if (id != null) {
+      linkAttributes.href = getFileUrl(id, 'full', fileName)
+      this.options.reportNode?.(id, node)
+    }
+    const icon = document.createElement('div')
+    icon.classList.add('icon')
+    icon.innerHTML = fileType.includes('image') === true ? imageIcon : attachIcon
+
+    return [
+      'div',
+      nodeAttributes,
+      ['div', { class: 'file-name-container' }, icon, ['a', linkAttributes, `${fileName}`]],
+      ['div', { class: 'file-size' }, `${filesize(size)}`]
+    ]
+  },
+
+  addInputRules () {
+    return [
+      nodeInputRule({
+        find: inputRegex,
+        type: this.type,
+        getAttributes: (match) => {
+          const [, , alt, src, title] = match
+
+          return { src, alt, title }
+        }
+      })
+    ]
+  },
+
+  addProseMirrorPlugins () {
+    const opt = this.options
+    function handleDrop (
+      view: EditorView,
+      pos: { pos: number, inside: number } | null,
+      dataTransfer: DataTransfer
+    ): any {
+      let result = false
+
+      const files = dataTransfer?.files
+      if (files !== undefined && opt.attachFile !== undefined) {
+        enum FileType {
+          none = 0,
+          image = 1,
+          file = 2,
+          mix = 3 // 1 + 2 = image + file
+        }
+        let attachTypes: FileType = 0
+        for (let i = 0; i < files.length; i++) {
+          const file = files.item(i)
+          if (file != null) {
+            if (file.type.includes('image')) {
+              attachTypes = attachTypes | FileType.image
```

Comment:
can be simplified to two boolean variables (hasImage, hasFile)

Reply:
Done